pkg/deploy/controller/deployerpod/controller.go

```diff
@@ -63,16 +75,78 @@ func (c *DeployerPodController) Handle(pod *kapi.Pod) error {
 	return nil
 }
 
+func (c *DeployerPodController) cleanupFailedDeployment(deployment *kapi.ReplicationController) error {
+	// Scale down the current failed deployment
+	configName := deployutil.DeploymentConfigNameFor(deployment)
+	existingDeployments, err := c.deploymentClient.listDeploymentsForConfig(deployment.Namespace, configName)
+	if err != nil {
+		return fmt.Errorf("couldn't list Deployments for DeploymentConfig %s: %v", configName, err)
+	}
+
+	desiredReplicas, success := deployutil.DeploymentDesiredReplicas(deployment)
+	if !success {
+		return fmt.Errorf("Could not determine desired replicas from %s to reset replicas for last completed deployment", deployutil.LabelForDeployment(deployment))
+	}
+
+	// construct the retry config to be used by the scaler for scaling up/down
+	// the completed and failed deployments
+	retryWaitParams := kubectl.NewRetryParams(1*time.Second, 120*time.Second)
+
+	if len(existingDeployments.Items) > 0 {
+		sort.Sort(deployutil.DeploymentsByLatestVersionDesc(existingDeployments.Items))
+		for index, existing := range existingDeployments.Items {
+			// if a newer deployment exists:
+			// - set the replicas for the current failed deployment to 0
+			// - there is no point in scaling up the last completed deployment
+			// since that will be scaled down by the later deployment
+			if index == 0 && existing.Name != deployment.Name {
+				break
+			}
+
+			// the first completed deployment is the one that needs to be scaled back up
+			if deployutil.DeploymentStatusFor(&existing) == deployapi.DeploymentStatusComplete {
+				if existing.Spec.Replicas == desiredReplicas {
+					break
+				}
+
+				// scale back the completed deployment to the target of the failed deployment
+				if err := c.scaler.Scale(existing.Namespace, existing.Name, uint(desiredReplicas), &kubectl.ScalePrecondition{-1, ""}, retryWaitParams, retryWaitParams); err != nil {
```

Comment:
My original code had one-shot updates - I changed it based on feedback. I can change it back, you want.